File: dbxml/examples/java/basic/Query.java
/*
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 2004,2009 Oracle.  All rights reserved.
 *
 */

package basic;

import com.sleepycat.dbxml.XmlException;
import com.sleepycat.dbxml.XmlManager;
import com.sleepycat.dbxml.XmlContainer;
import com.sleepycat.dbxml.XmlDocument;
import com.sleepycat.dbxml.XmlUpdateContext;
import com.sleepycat.dbxml.XmlQueryExpression;
import com.sleepycat.dbxml.XmlResults;
import com.sleepycat.dbxml.XmlValue;
import com.sleepycat.dbxml.XmlQueryContext;

/**
 * Query is a very simple Berkeley DB XML program
 * that performs a query and handles results.
 * It demonstrates initialization, container creation,
 * document insertion, query creation and execution,
 * use of a variable in a query and context, and
 * results handling
 * <p>
 * To run the example:
 * <pre>
 * java basic.Query
 * </pre>
 */
class Query {
    // This function is used to ensure that databases are
    // properly closed, even on exceptions
    private static void cleanup(XmlManager mgr, XmlContainer cont) {
	try {
	    if (cont != null)
		cont.delete();
	    if (mgr != null)
		mgr.delete();
	} catch (Exception e) {
	    // ignore exceptions in cleanup
	}
    }

    public static void main(String args[])
	throws Throwable {
	
	// This program uses a named container, which will apear
	// on disk
	String containerName = "people.dbxml";
	String content = "<people><person><name>joe</name></person><person><name>mary</name></person></people>";
	String docName = "people";
	// Note that the query uses a variable, which must be set
	// in the query context
	String queryString =
	    "collection('people.dbxml')/people/person[name=$name]";

	// declare these here for cleanup
	XmlManager mgr = null;
	XmlContainer cont = null;
	try {
	    // All BDB XML programs require an XmlManager instance
	    mgr = new XmlManager();

	    // Because the container will exist on disk, remove it
	    // first if it exists
	    if (mgr.existsContainer(containerName) != 0)
		mgr.removeContainer(containerName);

	    // Now it is safe to create the container
	    cont = mgr.createContainer(containerName);

	    cont.putDocument(docName, content);

	    // Querying requires an XmlQueryContext
	    XmlQueryContext qc = mgr.createQueryContext();
	    
	    // Add a variable to the query context, used by the query
	    qc.setVariableValue("name", new XmlValue("mary"));
	    
	    // Note: these two calls could be replaced by a shortcut
	    // mgr.query(queryString, qc), but under the covers,
	    // this is what is happening.
	    XmlQueryExpression expr = mgr.prepare(queryString, qc);
	    XmlResults res = expr.execute(qc);
	    
	    // Note use of XmlQueryExpression::getQuery() and
	    // XmlResults::size()
	    System.out.println("The query, '" + expr.getQuery() +
			       "'\n\t returned " + res.size() + " result(s)");
	    
	    // Process results -- just print them
	    XmlValue value = new XmlValue();
	    System.out.print("Result: ");
	    while ((value = res.next()) != null) {
		System.out.println("\t" + value.asString());
	    }

	    // Explicitly delete other objects to release resources
	    res.delete();
	    expr.delete();
	} catch (XmlException xe) {
	    System.err.println("XmlException during helloWorld: " + 
			       xe.getMessage());
	    throw xe;
	}
	finally {
	    cleanup(mgr, cont);
	}
    }
}

File: dbxml/examples/python/basic/query.py
"""
 * Query is a very simple Berkeley DB XML program
 * that performs a query and handles results.
 * It demonstrates initialization, container creation,
 * document insertion, query creation and execution,
 * use of a variable in a query and context, and
 * results handling.
 *
 * To run the example:
 *     python query.py
"""
from dbxml import *

def initialize(cname):
    # all BDB XML programs require an XmlManager instance
    mgr = XmlManager()
    
    # remove the container if it exists
    if mgr.existsContainer(cname) != 0:
        mgr.removeContainer(cname)

    # create a default container type
    cont = mgr.createContainer(cname)
    return cont

def query(mgr, cont):
    content = "<people><person><name>joe</name></person><person><name>mary</name></person></people>"
    docName = "people"
    queryString = "collection('%s')/people/person[name=$name]"%cont.getName()
    try:
        # all modification operations need XmlUpdateContext
        uc = mgr.createUpdateContext()
        cont.putDocument(docName, content, uc)
        
        # querying requires an XmlQueryContext
        qc = mgr.createQueryContext()

        # add a variable to the context, as required by the query
        qc.setVariableValue("name", XmlValue("mary"))

        # note: these two calls could be replaced by a shortcut
        # mgr.query(queryString, qc), but under the covers,
        # this is what is happening.
        expr = mgr.prepare(queryString, qc)
        res = expr.execute(qc)
	    
        # note use of XmlQueryExpression.getQuery() and
        # XmlResults.size()
        print "The query '%s'\n\t returned %d result(s)"%(expr.getQuery(),res.size())
	    
        # process results -- just print them.  XmlResults
        # are a Python iterator
	for value in res:
	    print "Result: %s"%value.asString()

    except XmlException, inst:
        print inst

import sys
if __name__ == "__main__":
    try:
        containerName = "people.dbxml"
        cont = initialize(containerName)
    except XmlException, xe:
        print xe
        sys.exit()

    query(cont.getManager(), cont)

